### PR TITLE
feat: Add automated pre-commit fix workflow for main branch and pull requests

### LINKED ISSUE #4438: Automate pre-Commit fixes for example if main has a pre-Commit issue then have it automatically create a new branch and pull across if a pull request has a pre-Commit issue automate request to that external branch

### PR DESCRIPTION
**Pull request by arnavkirti**
### 📝 Description
This PR introduces a GitHub Actions workflow that automatically detects and fixes pre-commit violations on the `main` branch, creating a PR with the corrections for review.

### ✨ Features
- **Automatic Detection**: Runs pre-commit checks on every push to main
- **Auto-Fix**: Applies Black, isort, ruff, and djLint formatting automatically
- **PR Creation**: Opens a PR with all fixes for maintainer review
- **Commit Comments**: Adds helpful comments to triggering commits
- **Workflow Summary**: Provides detailed execution summary in Actions UI

### 🔧 How It Works
1. Workflow triggers on push to `main` branch
2. Runs `pre-commit run --all-files`
3. If violations are detected and auto-fixable:
   - Creates a new branch: `fix/pre-commit-main-YYYYMMDD-HHMMSS`
   - Commits the auto-fixes
   - Opens a PR targeting `main`
   - Adds a comment to the original commit with PR link

### 🧪 Testing
Tested thoroughly on `test/*` branches with intentional formatting violations:
- ✅ Successfully detected violations
- ✅ Applied auto-fixes correctly
- ✅ Created fix branches
- ✅ Opened PRs with proper formatting
- ✅ Handled edge cases (no fixes needed, permission issues)

### 📸 Screenshots
Screenshot of successful workflow run:
Screenshot 2025-12-03 at 6 52 55 PM

Screenshot of successful PR creation:
Screenshot 2025-12-03 at 6 49 56 PM

### 🎯 Benefits
- **Reduces Manual Work**: No need to manually fix formatting issues
- **Maintains Code Quality**: Ensures consistent formatting on main
- **Transparent**: All fixes are reviewable before merging
- **Non-Disruptive**: Works alongside existing CI/CD pipelines

### ⚠️ Notes
- Labels (`automated-fix`, `pre-commit`) need to be added manually to PRs due to token permissions
- Workflow only triggers on `main` branch pushes
- Requires `contents: write` and `pull-requests: write` permissions (already configured)

### 📚 Related Issues
Resolves #4438 

### ✅ Checklist
- [x] Tested on test branches
- [x] Workflow creates PRs successfully
- [x] Handles edge cases gracefully
- [x] Follows project coding standards
- [x] Workflow includes proper error handling

## Summary by CodeRabbit

* **Chores**
  * Added automated workflows that run pre-commit checks on main and on pull requests; they can apply fixes, create or update fix branches/PRs, add labels and comments, throttle activity and avoid duplicate runs, enforce diff-size limits, handle retries and remote conflicts, cache dependencies, and emit clear success/failure summaries with guidance.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by arnavkirti** [comment]
@Nachiket-Roy Everything has been resolved now. A review would be very helpful.
Thanks.

**Comment by arnavkirti** [comment]
@sidd190 I resolved the null-safety checks. Thanks.

**Review by Nachiket-Roy — CHANGES_REQUESTED** [review]
There is a Potential Race condition in fetch and check for new commits (add a retry  mechanisms).

File: .github/workflows/auto-fix-main-precommit.yml [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+          ### 📋 Files Changed:
+          _(Unable to retrieve changed files list. Check the commit diff in the PR for details.)_
+          EOF
+          fi
+
+          cat >> "$PR_BODY_FILE" <<EOF
+
+          ### 🏷️ Labels
+          Please manually add these labels: \`automated-fix\`, \`pre-commit\`
```
Comment (Nachiket-Roy) [inline .github/workflows/auto-fix-main-precommit.yml]: The workflow currently asks contributors to manually add the labels automated-fix and pre-commit, but this can be fully automated.

**Comment by arnavkirti** [comment]
Thank you for the review @Nachiket-Roy, I have implemented all the changes you suggested.

Summary of Changes

This workflow now safely automates pre-commit fixes on PRs while handling concurrency and avoiding branch conflicts:
	•	Added per-branch concurrency control so only one auto-fix run executes at a time.
	•	Ensures auto-fixes only run on same-repo PRs, never fork branches.
	•	Detects whether pre-commit produced changes and only attempts auto-fix when applicable.
	•	Added conflict detection — if the PR branch changes during the run, the workflow backs off and comments instead of pushing.
	•	Removed CI rebase logic to avoid rewriting contributor history.
	•	Uses git add -u to commit only tracked file changes.
	•	After pushing fixes, the workflow re-runs pre-commit to verify success; fails the job if issues remain.
	•	Improved PR comments and summaries for clear communication of auto-fix status.

This results in a safe, predictable, and contributor-friendly auto-fix workflow.

**Comment by Nachiket-Roy** [comment]
I also suggest testing this thoroughly before merging. @coderabbitai Could you recommend some ways to test it?

**Comment by Nachiket-Roy** [comment]
@arnavkirti something the workflow hangs for too long and requires re-run we should add a retry or fail it after sometime. what do you think?

**Comment by arnavkirti** [comment]
@Nachiket-Roy, Thanks for pointing that out. yes, there are a few steps (Poetry install, pre-commit run, or network-bound gh calls) that can occasionally hang longer than expected.

Right now we already have retry logic around the push + conflict-detection steps, but the install + pre-commit phases don’t have a retry or timeout guard.

We can improve this in two ways:
	1.	Add a global job-level timeout (e.g., timeout-minutes: 10) so the workflow fails fast instead of hanging indefinitely.
	2.	Add targeted retry wrappers around the steps most likely to hang (Poetry install, pre-commit run) to make transient failures self-heal.
	
I’ll add both:
	•	a timeout-minutes at the job level
	•	a retry wrapper for the dependency install + pre-commit steps

This should make the workflow much more resilient and avoid cases where it gets stuck.

Do you think this approach is correct??

**Comment by arnavkirti** [comment]
@Nachiket-Roy, can you please review this again? I have implemented all the changes necessary, and now the workflows are merge-ready.

**Review by Nachiket-Roy — APPROVED** [review]

**Comment by arnavkirti** [comment]
Screenshot 2025-12-06 at 1 09 32 PM
<br>

I have tested everything extensively using `act` locally. All of the tests are passing.
